readme.md
[![Commitizen friendly](https://img.shields.io/badge/commitizen-friendly-brightgreen.svg)](http://commitizen.github.io/cz-cli/)

# Ant Party

## dev

This project relies on a redis server. When running the scripts below, a script will be called that runs a redis Docker image.
To start the dev server (client & backend), run:

```bash
npm i
npm run dev
```

This runs the dev server on port `8000` and you can view it here:

http://localhost:8000/api/v1/test/

...and then seperately run the client in the `./client` subdirectory:

```bash
cd ./client
npm i
npm run start
```

http://localhost:3000/ should now host the app, and suport hot reloading.

## contributing

Hopefully this project can remain [Commitizen friendly](http://commitizen.github.io/cz-cli/).

In short, to commit, run:

```bash
npx git-cz
```

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.3.8](https://github.com/jonpepler/ant-party/compare/v0.3.7...v0.3.8) (2021-01-15)


### Features

* 🎸 Don't move ants onto occupied spaces ([c26a517](https://github.com/jonpepler/ant-party/commit/c26a51716b4ff259f4a7c7dcc6b0626b7e35afc8))
* 🎸 Support pheromones ([33f2162](https://github.com/jonpepler/ant-party/commit/33f216260da58c626af1b2eb409b81a8548eef8c))


### Bug Fixes

* 🐛 Increase server ping frequency to help connection drops ([b740a7b](https://github.com/jonpepler/ant-party/commit/b740a7b6704343bdc144b86f819b1c7902504eaf))

### [0.3.7](https://github.com/jonpepler/ant-party/compare/v0.3.6...v0.3.7) (2021-01-11)


### Features

* 🎸 Only spawn ants when requested by a player ([ccaae9a](https://github.com/jonpepler/ant-party/commit/ccaae9a6e466d1f7b3429ccc521e6146a075eda1))
* 🎸 Spawn ants in at one of four free spawn points on nests ([2531b5a](https://github.com/jonpepler/ant-party/commit/2531b5ad384f0cadbe7a06a5533c9f28f09e5b87))

### [0.3.6](https://github.com/jonpepler/ant-party/compare/v0.3.5...v0.3.6) (2021-01-08)


### Bug Fixes

* 🐛 Set node-sass back to 4.x.x ([9eeb36f](https://github.com/jonpepler/ant-party/commit/9eeb36f83a74b092dda1e70709f2200e06a918e3))

### [0.3.5](https://github.com/jonpepler/ant-party/compare/v0.3.4...v0.3.5) (2021-01-08)


### Bug Fixes

* 🐛 Remove test reference to GameSketch ([112740b](https://github.com/jonpepler/ant-party/commit/112740bb9614b13f763c37e1bdeb7a8c3072d0e2))

### [0.3.4](https://github.com/jonpepler/ant-party/compare/v0.3.3...v0.3.4) (2021-01-08)

### [0.3.3](https://github.com/jonpepler/ant-party/compare/v0.3.2...v0.3.3) (2021-01-08)


### Features

* 🎸 Display ants and nests based on server data ([0878f8c](https://github.com/jonpepler/ant-party/commit/0878f8c5cfd03753180b8004ecbb204e62bef755))

### [0.3.2](https://github.com/jonpepler/ant-party/compare/v0.3.1...v0.3.2) (2020-07-18)


### Bug Fixes

* 🐛 Re-add button border ([48d1c02](https://github.com/jonpepler/ant-party/commit/48d1c020efbc84b21cfa535baa4cada8c984bde4))

### [0.3.1](https://github.com/jonpepler/ant-party/compare/v0.3.0...v0.3.1) (2020-03-19)

## [0.3.0](https://github.com/jonpepler/ant-party/compare/v0.2.1...v0.3.0) (2020-03-19)


### Features

* 🎸 Delete player record when they quit ([172b32f](https://github.com/jonpepler/ant-party/commit/172b32feff92175a3dde8bc71e15883d75397c6b))
* 🎸 Only allow game admin from the game host ([12cd6cd](https://github.com/jonpepler/ant-party/commit/12cd6cd0228873ad1c1c78910b60f5c6d714054c))
* 🎸 Reflect submitted ant.js files back to the players ([4bfa470](https://github.com/jonpepler/ant-party/commit/4bfa470f1a2598ec3c2588064339c054066f6713))
* 🎸 Show a start button when enough players have joined ([4034ab8](https://github.com/jonpepler/ant-party/commit/4034ab8cf94869e480bcf108de35a97ed0c0abfa))
* 🎸 Show a very basic p5 canvas on game start ([a8fa6dd](https://github.com/jonpepler/ant-party/commit/a8fa6ddf04de46035e3186a2ab7e4396ff273431))
* 🎸 Show basic ants and dirt in scalable P5 demo ([816ab5e](https://github.com/jonpepler/ant-party/commit/816ab5e87a3fa78ed4659e6833ddfc80cf7c3a3b))
* 🎸 Show connection error in footer when con fails happen ([8e2e666](https://github.com/jonpepler/ant-party/commit/8e2e6667d51906591de96168da628f8329428859))
* 🎸 Start game when instructed by host ([2df4b93](https://github.com/jonpepler/ant-party/commit/2df4b9312edf77d54bfa83a509d23a746f814271))


### Bug Fixes

* 🐛 Add missing dependencies (p5/react-p5) ([5743e15](https://github.com/jonpepler/ant-party/commit/5743e155e773491196a2cfd3af47bbe09f08767e))

### [0.2.1](https://github.com/jonpepler/ant-party/compare/v0.2.0...v0.2.1) (2020-03-08)

## [0.2.0](https://github.com/jonpepler/ant-party/compare/v0.1.2...v0.2.0) (2020-03-08)


### Features

* 🎸 Show player IDs in web client when they join ([f5eef4c](https://github.com/jonpepler/ant-party/commit/f5eef4ce61b4eaf2c770fd7bac10ee3df935f4fa))
* 🎸 Show player names when they join a game ([1acd87b](https://github.com/jonpepler/ant-party/commit/1acd87bac06647f1a4989d7601474edcd95e49d6))
* 🎸 Track when players leave and show in the web view ([b732404](https://github.com/jonpepler/ant-party/commit/b73240455720d7e6899c3641b4bbc59d9bfb25f4))


### Bug Fixes

* 🐛 Fix redis dev script attempting to run the application ([47517a3](https://github.com/jonpepler/ant-party/commit/47517a3f112e464c028f371cf107440aaf578d35))

### [0.1.2](https://github.com/jonpepler/ant-party/compare/v0.1.1...v0.1.2) (2020-03-07)

### [0.1.1](https://github.com/jonpepler/ant-party/compare/v0.0.3...v0.1.1) (2020-03-07)


### Features

* 🎸 Receive game join requests over socket ([379c4c7](https://github.com/jonpepler/ant-party/commit/379c4c7d713092c062bacdebab7e256e9c42555b))
* 🎸 Track game events via socket tracker in client ([065a5df](https://github.com/jonpepler/ant-party/commit/065a5df2a56980f194db6da48b29c2871fe1fc21))

### [0.0.3](https://github.com/jonpepler/ant-party/compare/v0.0.2...v0.0.3) (2020-03-06)


### Features

* 🎸 Add a basic start game flow ([6726e4f](https://github.com/jonpepler/ant-party/commit/6726e4f7f9562e818bef5d2d3733cb03dd10ffe1))
* 🎸 Add a hover effect on buttons ([795c401](https://github.com/jonpepler/ant-party/commit/795c4010788c21c319744ac5ca962a1b25c9f5fc))
* 🎸 Allow players to join a game via ant-party-client ([3147632](https://github.com/jonpepler/ant-party/commit/3147632198ce82d4c74e1ba26ca7c50ec6322891))
* 🎸 Return a game code when starting a game ([df23a60](https://github.com/jonpepler/ant-party/commit/df23a60ea5563eba75cc7ee145cf24daff360271))
* 🎸 Update all terminology to refer to Games and gamecodes ([a85df7e](https://github.com/jonpepler/ant-party/commit/a85df7ed4731a92294c32cbad403352f1fe2c793))

app.js
const express = require('express')
const bodyParser = require('body-parser')
const { resolve } = require('path')
const history = require('connect-history-api-fallback')
const cors = require('cors')
const publicPath = resolve(__dirname, 'client/build')
const staticConf = { maxAge: '1y', etag: false }

const { session, redis, RedisStore } = require('./redis')

class App {
  constructor (controllers) {
    this.app = express()

    this.initializeMiddlewares()
    this.initializeControllers(controllers)
    this.initializeFrontEndMiddleware()
  }

  setupCors () {
  // These settings need to come from the environment
    const whitelist = [
      'http://localhost', // Local
      'http://localhost:8000', // Backend
      'http://localhost:3000', // Client (standard)
      'https://ant-party.herokuapp.com', // Demo
      'http://ant-party.herokuapp.com' // Demo
    ]
    const corsOptions = {
      origin: function (origin, callback) {
        if (!origin || (origin && whitelist.indexOf(origin) !== -1)) {
          callback(null, true)
        } else {
          callback(new Error('Not allowed by CORS'))
        }
      }
    }
    return cors(corsOptions)
  }

  initializeMiddlewares () {
    redis.on('error', (err) => {
      console.log('Redis error: ', err)
    })

    // Start a session; we use Redis for the session store.
    // "secret" will be used to create the session ID hash (the cookie id and the redis key value)
    // "name" will show up as your cookie name in the browser
    // "cookie" is provided by default; you can add it to add additional personalized options
    // The "store" ttl is the expiration time for each Redis session ID, in seconds
    this.app.use(session({
      secret: process.env.SECRET,
      name: '_redis',
      resave: false,
      saveUninitialized: true,
      cookie: { secure: process.env.NODE_ENV === 'production' },
      store: new RedisStore({ client: redis, ttl: 86400 })
    }))

    this.app.use(bodyParser.json())
    this.app.use((req, res, next) => {
      req.params.id = req.params.id ? Number(req.params.id) : undefined
      next()
    })
    this.app.use(this.setupCors())
  }

  initializeFrontEndMiddleware () {
    this.app.use('/', history()) // The history middleware enables the front-end to be served correctly by the back-end
    this.app.use(express.static(publicPath, staticConf))
    this.app.use('/', history()) // This has to be initialised twice for reloading of front-end routes to be handled correctly
  }

  initializeControllers (controllers) {
    controllers.forEach((controller) => {
      this.app.use('/api/v1/', controller.router)
    })
  }
}

module.exports = App
